File: sky/packages/sky/lib/src/rendering/paragraph.dart
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:sky/src/painting/text_painter.dart';
import 'package:sky/src/rendering/box.dart';
import 'package:sky/src/rendering/object.dart';

export 'package:sky/src/painting/text_painter.dart';

/// A render object that displays a paragraph of text
class RenderParagraph extends RenderBox {

  RenderParagraph(
    TextSpan text
  ) : textPainter = new TextPainter(text) {
    assert(text != null);
  }

  final TextPainter textPainter;

  BoxConstraints _constraintsForCurrentLayout; // when null, we don't have a current layout

  /// The text to display
  TextSpan get text => textPainter.text;
  void set text(TextSpan value) {
    if (textPainter.text == value)
      return;
    textPainter.text = value;
    _constraintsForCurrentLayout = null;
    markNeedsLayout();
  }

  // Whether the text should be allowed to wrap to multiple lines.
  bool get allowLineWrap => true;

  void layoutText(BoxConstraints constraints) {
    assert(constraints != null);
    if (_constraintsForCurrentLayout == constraints)
      return; // already cached this layout
    textPainter.maxWidth = allowLineWrap ? constraints.maxWidth : double.INFINITY;
    textPainter.minWidth = constraints.minWidth;
    textPainter.minHeight = constraints.minHeight;
    textPainter.maxHeight = constraints.maxHeight;
    textPainter.layout();
    _constraintsForCurrentLayout = constraints;
  }

  double getMinIntrinsicWidth(BoxConstraints constraints) {
    layoutText(constraints);
    return constraints.constrainWidth(textPainter.minContentWidth);
  }

  double getMaxIntrinsicWidth(BoxConstraints constraints) {
    layoutText(constraints);
    return constraints.constrainWidth(textPainter.maxContentWidth);
  }

  double _getIntrinsicHeight(BoxConstraints constraints) {
    layoutText(constraints);
    return constraints.constrainHeight(textPainter.height);
  }

  double getMinIntrinsicHeight(BoxConstraints constraints) {
    return _getIntrinsicHeight(constraints);
  }

  double getMaxIntrinsicHeight(BoxConstraints constraints) {
    return _getIntrinsicHeight(constraints);
  }

  double computeDistanceToActualBaseline(TextBaseline baseline) {
    assert(!needsLayout);
    layoutText(constraints);
    return textPainter.computeDistanceToActualBaseline(baseline);
  }

  void performLayout() {
    layoutText(constraints);

    // We use textPainter.maxContentWidth here, rather that textPainter.width,
    // because the latter is the width that it used to wrap the text, whereas
    // the former is the actual width of the text.
    size = constraints.constrain(new Size(textPainter.maxContentWidth,
                                          textPainter.height));
  }

  void paint(PaintingContext context, Offset offset) {
    // Ideally we could compute the min/max intrinsic width/height with a
    // non-destructive operation. However, currently, computing these values
    // will destroy state inside the painter. If that happens, we need to
    // get back the correct state by calling _layout again.
    //
    // TODO(abarth): Make computing the min/max intrinsic width/height
    // a non-destructive operation.
    layoutText(constraints);
    textPainter.paint(context.canvas, offset);
  }

  // we should probably expose a way to do precise (inter-glpyh) hit testing

  String debugDescribeSettings(String prefix) {
    String result = '${super.debugDescribeSettings(prefix)}';
    result += '${prefix}text:\n${text.toString("$prefix  ")}\n';
    return result;
  }
}

File: impeller/toolkit/interop/paragraph.cc
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "impeller/toolkit/interop/paragraph.h"

namespace impeller::interop {

Paragraph::Paragraph(std::unique_ptr<txt::Paragraph> paragraph)
    : paragraph_(std::move(paragraph)) {}

Paragraph::~Paragraph() = default;

Scalar Paragraph::GetMaxWidth() const {
  return paragraph_->GetMaxWidth();
}

Scalar Paragraph::GetHeight() const {
  return paragraph_->GetHeight();
}

Scalar Paragraph::GetLongestLineWidth() const {
  return paragraph_->GetLongestLine();
}

Scalar Paragraph::GetMinIntrinsicWidth() const {
  return paragraph_->GetMinIntrinsicWidth();
}

Scalar Paragraph::GetMaxIntrinsicWidth() const {
  return paragraph_->GetMaxIntrinsicWidth();
}

Scalar Paragraph::GetIdeographicBaseline() const {
  return paragraph_->GetIdeographicBaseline();
}

Scalar Paragraph::GetAlphabeticBaseline() const {
  return paragraph_->GetAlphabeticBaseline();
}

uint32_t Paragraph::GetLineCount() const {
  return paragraph_->GetNumberOfLines();
}

const std::unique_ptr<txt::Paragraph>& Paragraph::GetHandle() const {
  return paragraph_;
}

}  // namespace impeller::interop
